zip-0316.rst

```diff
@@ -554,6 +555,53 @@ introduced either by a modification to this ZIP or by a new ZIP, in
 accordance with the ZIP Process [#zip-0000]_.
 
 
+Deriving Internal Keys
+----------------------
+
+In addition to external addresses suitable for giving out to Senders,
+a wallet typically requires addresses for internal operations such as
+change and auto-shielding.
+
+We desire the following properties for viewing authority of both
+shielded and transparent key trees:
+
+- A holder of an FVK can derive external and internal IVKs, and
+  external and internal :math:`\mathsf{ovk}` components.
+
+- A holder of the external IVK cannot derive the internal IVK, or
+  any of the :math:`\mathsf{ovk}` components.
+
+- A holder of the external :math:`\mathsf{ovk}` component cannot derive
+  the internal :math:`\mathsf{ovk}` component, or any of the IVKs.
+
+For shielded keys, these properties are achieved by the one-wayness of
+:math:`\mathsf{PRF^{expand}}` and of :math:`\mathsf{CRH^{ivk}}` or
+:math:`\mathsf{Commit^{ivk}}` (for Sapling and Orchard respectively).
+Derivation of an internal shielded FVK from an external shielded FVK
+is specified in the
+"Sapling internal key derivation" [#zip-0032-sapling-internal-key-derivation]_ and
+"Orchard internal key derivation" [#zip-0032-orchard-internal-key-derivation]_
+sections of ZIP 32.
+
+To satisfy the above properties for transparent (P2PKH) keys, we derive
```

Comment:
A unified spending key is not specific to P2PKH, those are designations for addresses. However, for a P2SH spend, we still need to be able to define an OVK for shielding, right? Do we just use the same OVK as is produced for shielding from the P2PKH address?

Reply:
There needs to be a section in ZIP 316 that discusses usage of OVKs.

Reply:
There should be a section of ZIP 316 that documents their intended usage semantics, so that other wallets implement them consistently.

The current `zcashd` and mobile SDK behaviour is:
- If a transaction spends coins/notes from more than one spend authority, use the OVK for the first coin/note added to the transaction builder.
  - This ends up being implementation-specific behaviour that depends on the coin selection algorithm, but the effect is that a single OVK is used for all outputs, corresponding to only one of the input spend authorities.
  - The only current case where there could be more than one spend authority is `z_mergetoaddress`.

What we need to specify in ZIP 316 is the UA-internal OVK selection logic. In our pairing today, @therealyingtong and I worked through how OVKs would need to be used, and came up with a few ideas. In ZIP sync, we came up with the following proposal:
- Use the external and internal OVKs derived from the most preferred typecode in the account's UFVK, out of the typecodes coresponding to coins/notes being spent in the transaction.